src/libraries/System.Private.CoreLib/src/System/Runtime/CompilerServices/RuntimeHelpers.cs

```diff
@@ -208,6 +208,75 @@ public static void UnsafeAwaitAwaiterFromRuntimeAsync<TAwaiter>(TAwaiter awaiter
             SuspendAsync2(sentinelContinuation);
             return;
         }
+
+        // TODO: should this be called "AwaitFromRuntimeAsync" ?  (i.e. same as above, but no "Awaiter")
+        //
+        // Marked intrinsic since this needs to be
+        // recognizes as an async2 call.
+        [Intrinsic]
+        [BypassReadyToRun]
+        [MethodImpl(MethodImplOptions.NoInlining)]
+        public static unsafe T Await<T>(Task<T> task)
+        {
+            // TODO: handle complete tasks more efficiently.
+            //if (!task.IsCompleted)
+            //{
+                ref RuntimeAsyncAwaitState state = ref t_runtimeAsyncAwaitState;
+                Continuation? sentinelContinuation = state.SentinelContinuation;
+                if (sentinelContinuation == null)
+                    state.SentinelContinuation = sentinelContinuation = new Continuation();
+
+                Continuation myContinuation = new Continuation();
+                myContinuation.GCData = new object[] { task };
+                myContinuation.Resume = &AwaitHelper<T>.Resume;
+
+                state.Notifier = task.GetAwaiter();
+                sentinelContinuation.Next = myContinuation;
+
+                //     RETURN {default(T), myContinuation}
+                //
+                SuspendAsync2(myContinuation);
+
+                // unreachable
+                return task.ResultOnSuccess;
+            //}
+            //else
+            //{
+            //    //   RETURN {task.Result, null}
+            //    //
+            //    T result = task.Result;
+            //    ReturnAsync2(Unsafe.AsPointer(ref result));
+            //
+            //    // unreachable
+            //    return result;
+            //}
+        }
+
+        internal static class AwaitHelper<T>
+        {
```

Comment:
Resuming the continuation simply pushes the result into the caller continuation. (or throws, if faulted)

Reply:
I wonder if `UnsafeAwaitAwaiterFromRuntimeAsync` needs to call `GetResult()` upon resuming as well? (in case it faulted)

Reply:
Also `GetAwaiter().GetResult()` seems more appropriate than `.Result`. This is just a rough prototype anyways.

Reply:
> I wonder if `UnsafeAwaitAwaiterFromRuntimeAsync` needs to call `GetResult()` upon resuming as well? (in case it faulted)

The caller is responsible for calling `GetAwaiter().GetResult()` (`UnsafeAwaitAwaiterFromRuntimeAsync` has no way to bind to the awaiter anyway).

Reply:
> The caller is responsible for calling awaiter.GetResult()

Ah, right, in this case we have code in the caller.